signalling/src/server/mod.rs

```diff
@@ -0,0 +1,227 @@
+use anyhow::Error;
+use async_std::net::{TcpListener, TcpStream};
+use async_std::task;
+use async_tungstenite::tungstenite::Message as WsMessage;
+use futures::channel::mpsc;
+use futures::prelude::*;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use tracing::{info, trace, warn};
+
+use crate::handlers::{MessageHandler, MessageSender};
+use crate::protocol as p;
+
+#[derive(Clone)]
+/// The extendable signalling server
+pub struct SignallingServer {
+    state: Arc<Mutex<State>>,
+}
+
+struct Peer {
+    receive_task_handle: task::JoinHandle<()>,
+    send_task_handle: task::JoinHandle<Result<(), Error>>,
+    sender: mpsc::Sender<String>,
+}
+
+#[derive(Default)]
+struct DefaultMessageSender {
+    peers: HashMap<String, Peer>,
+}
+
+impl MessageSender for DefaultMessageSender {
+    fn send_message(&mut self, message: String, peer_id: &str) {
+        if let Some(peer) = self.peers.get_mut(peer_id) {
+            let mut sender = peer.sender.clone();
+            let peer_id = peer_id.to_string();
+            task::spawn(async move {
+                if let Err(err) = sender.send(message).await {
+                    warn!(peer_id = %peer_id, "Failed to send message: {}", err);
+                }
+            });
+        }
+    }
+}
+
+struct State {
+    message_handler: Box<dyn MessageHandler>,
+    message_sender: Option<DefaultMessageSender>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum SignallingServerError {}
+
+impl SignallingServer {
+    /// Instantiate the signalling server with a MessageHandler
+    /// implementation. Use DefaultMessageHandler for the default
+    /// protocol, or implement your own handler
+    pub fn new(message_handler: Box<dyn MessageHandler>) -> Self {
```

Comment:
```suggestion
    pub fn new<H: MessageHandler>(message_handler: H) -> Self {
        let message_handler = Box::new(message_handler);
```

That's a bit nicer to use